include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+      /* copy construct in place */
+      allocator_.construct(data_ + index, value);
+    }
+  }
+}
+
+template <class Type, class Allocator>
+template <class InputIt>
+void ekuvector<Type, Allocator>::assign(
+    InputIt first,
+    typename std::enable_if<is_iterator<InputIt>::value, InputIt>::type last) {
+  /* call a copy constructor on each */
+  size_t index = 0;
+  for (auto it = first; it != last; ++it, ++index) {
+    /* make sure we have enough capacity */
+    if (capacity_ < index + 1) {
+      preallocate_capacity(index + 1);
+    }
+    /* if the element was previously constructed, destroy  it first */
+    if (index < size_) {
+      allocator_.destroy(data_ + index);
+    } else {
+      ++size_;
+    }
+    /* copy construct in place */
+    allocator_.construct(data_ + index, *it);
+  }
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::assign(std::initializer_list<Type> ilist) {
+  /* call a copy constructor on each */
+  size_t index = 0;
+  for (auto it = ilist.begin(); it != ilist.end(); ++it, ++index) {
+    /* make sure we have enough capacity */
+    if (capacity_ < index + 1) {
+      preallocate_capacity(index + 1);
+    }
+    /* if the element was previously constructed, destroy first */
+    if (index < size_) {
+      allocator_.destroy(data_ + index);
+    } else {
+      ++size_;
+    }
+    /* copy construct in place */
+    allocator_.construct(data_ + index, *it);
+  }
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::allocator_type
+ekuvector<Type, Allocator>::get_allocator() const {
+  return allocator_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reference
+ekuvector<Type, Allocator>::at(size_type pos) {
+  if (pos >= size_) {
+    throw std::out_of_range("vector index out of range");
+  }
+  return *(data_ + pos);
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reference
+ekuvector<Type, Allocator>::at(size_type pos) const {
+  if (pos >= size_) {
+    throw std::out_of_range("vector index out of range");
+  }
+  return *(data_ + pos);
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reference ekuvector<Type, Allocator>::
+operator[](size_type pos) {
+  return *(data_ + pos);
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reference
+    ekuvector<Type, Allocator>::operator[](size_type pos) const {
+  return *(data_ + pos);
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reference
+ekuvector<Type, Allocator>::front() {
+  return *data_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::const_reference
+ekuvector<Type, Allocator>::front() const {
+  return *data_;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::reference
+ekuvector<Type, Allocator>::back() {
+  return (*this)[size_ - 1];
```

Comment:
I know it is undefined, but would nullptr in case of size() == 0 be better?

Reply:
SInce the API requires me to return a reference, it's not possible to return a nullpointer for that particular case (not without some wild casting to create a reference to the address pointed by the null pointer).

Reply:
Oh, I completely missed it! Thanks.